workflows/coproid.nf

```diff
@@ -0,0 +1,292 @@
+    //
+
+    PREPARE_GENOMES (
+        ch_genomesheet
+    )
+    ch_versions = ch_versions.mix(PREPARE_GENOMES.out.versions.first())
+
+    //
+    // MODULE: Run FastQC
+    //
+    FASTQC (
+        ch_samplesheet
+    )
+    ch_multiqc_files = ch_multiqc_files.mix(FASTQC.out.zip.collect{it[1]})
+    ch_versions = ch_versions.mix(FASTQC.out.versions.first())
+
+    //
+    // MODULE: Preprocessing with fastp
+    //
+    FASTP (
+        ch_samplesheet,
+        [],
+        false,
+        false,
+        true
+    )
+    ch_trimmed = FASTP.out.reads
+    ch_multiqc_files = ch_multiqc_files.mix(FASTP.out.json.collect{it[1]})
+    ch_versions = ch_versions.mix(FASTP.out.versions.first())
+
+    //
+    // SUBWORKFLOW: Align reads to all genomes and index alignments
+    //
+
+    FASTP.out.reads // [meta[ID, single_end], merged_reads]
+        .combine(PREPARE_GENOMES.out.genomes) //[meta[genome_name], fasta, index]
+        .map {
+            meta_reads, reads, meta_genome, genome_fasta, genome_index ->
+            [
+                [
+                    'id': meta_reads.id + '-' + meta_genome.genome_name,
+                    'genome_name': meta_genome.genome_name,
+                    'genome_taxid': meta_genome.taxid,
+                    'genome_size': meta_genome.genome_size,
+                    'sample_name': meta_reads.id,
+                    'single_end': meta_reads.single_end,
+                    'merge': meta_reads.merge
+                ],
+                reads,
+                genome_index,
+                genome_fasta
+            ]
+        }
+        .set { ch_reads_genomes_index }
+
+    ALIGN_INDEX (
+        ch_reads_genomes_index
+    )
+    ch_versions = ch_versions.mix(ALIGN_INDEX.out.versions.first())
+    ch_multiqc_files = ch_multiqc_files.mix(ALIGN_INDEX.out.multiqc_files.collect{it[1]})
+
+    DAMAGEPROFILER(
+        ALIGN_INDEX.out.bam,
+        [],
+        [],
+        []
+    )
+    ch_versions = ch_versions.mix(DAMAGEPROFILER.out.versions.first())
+    ch_multiqc_files = ch_multiqc_files.mix(DAMAGEPROFILER.out.results.collect{it[1]})
+
+    DAMAGEPROFILER.out.results.collect({it[1]})
+    .set { damageprofiler_reports }
+
+    DAMAGEPROFILER_MERGE(
+        damageprofiler_reports
+    )
+
+    ALIGN_INDEX.out.bam.join(
+        ALIGN_INDEX.out.bai
+    ).map {
+        meta, bam, bai -> [['id':meta.sample_name, 'genome_name':meta.genome_name], bam, bai] // meta.id, bam
+    }
+    .set { aligned_index }
+
+    PYDAMAGE_ANALYZE (
+        aligned_index
+    )
+    ch_versions = ch_versions.mix(PYDAMAGE_ANALYZE.out.versions.first())
+
+    PYDAMAGE_ANALYZE.out.csv.collect({it[1]})
+    .set { pydamage_reports }
+
+    PYDAMAGE_MERGE (
+        pydamage_reports
+    )
+
+    // join bam with indices
+    ALIGN_INDEX.out.bam.join(
+        ALIGN_INDEX.out.bai
+    ).map {
+        meta, bam, bai -> [['id':meta.sample_name], bam] // meta.id, bam
+    }.groupTuple()
```

Comment:
This will create a bottleneck in your pipeline. The pipeline will wait until all FASTQ files are aligned before actually performing the `.groupTuple` (which could cause significant slowdowns). 

You can fix this by telling the operator how big the expected groups should be. You can do this in two ways:
1. All groups will be the same size? Use the `size` option for the [operator](https://www.nextflow.io/docs/latest/reference/operator.html)
2. Group sizes are dynamic? Use the `groupKey` function to dynamically assign a group size per group key (This can also be found in the [.groupTuple](https://www.nextflow.io/docs/latest/reference/operator.html) docs

Reply:
Thanks for this @nvnieuwk! I think I have fixed that now with the size option. 
I'm not sure if I've done it the right way, but I made all the changes here https://github.com/nf-core/coproid/pull/79
I don't have direct writing access to dev.

Reply:
I approved the PR :) I'm not familiar with this pipeline so I don't know what the best solution is here. 

In your implementation, you told the pipeline that all channel entries after the `.groupTuple` should consist of the merge of two channel entries with the same ID.

Reply:
I don't know if the bottleneck is a big issue, as we anyway generate a single report per run, where the results of all samples are needed.

Reply:
@maxibor true. I guess the size operator will at least stop the bottlenecking before the quarto reporting? 

I changed it to .groupTuple(size: 2) - each tuple will contain the bam and the bai file, so I think this is right. I tested it with the test data, and it ran smoothly. I that looks good to you, I think we can merge https://github.com/nf-core/coproid/pull/79 ?